File: java/symmetric_tree_101/Solution.java
package symmetric_tree_101;

import tree.binary_tree.TreeNode;

public class Solution {
  public boolean isSymmetric(TreeNode root) {
    if (root == null) {
      return true;
    }

    return isMirror(root.left, root.right);
  }

  private static boolean isMirror(TreeNode left, TreeNode right) {
    if (left == null && right == null) {
      return true;
    } else if (left == null || right == null) {
      return false;
    }

    return isMirror(left.left, right.right)
        && isMirror(left.right, right.left)
        && left.val == right.val;
  }
}

File: python/symmetric_tree_101/Solution.py
from typing import Optional

from tree.binary_tree.TreeNode import TreeNode


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True

        return Solution.__isMirror(root.left, root.right)

    @staticmethod
    def __isMirror(left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
        if not left and not right:
            return True
        elif not left or not right:
            return False

        return (
            Solution.__isMirror(left.left, right.right)
            and Solution.__isMirror(left.right, right.left)
            and left.val == right.val
        )
